Fix errors and warnings building swift/SILGen on Windows using MSVC

hughbe opened the pull request:
- Reference deleted function error
- Constructor overload error
- Iterator internal stdlib error
- Disable assertion failing on MSVC

hughbe commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -174,7 +174,10 @@ class Callee {
 
   // Move, don't copy.
+  // Work around MSVC error: attempting to reference a deleted function.
+#if !defined(_MSC_VER) || defined(__clang__)
```
Note: replacing `delete` with `default` below fixes any errors (@jrose-apply)

hughbe replied:
Yes it does @jckarter. Thanks!!

hughbe replied:
You might want to take a look at https://github.com/apple/swift/pull/5948 in which I try to work around a similar error. Unfortunately, this time the code that triggers this is compiler generated `__delDtor(unsigned int)`

jckarter commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1693,7 +1693,7 @@ RValue RValueEmitter::visitMemberRefExpr(MemberRefExpr *E, SGFContext C) {
   if (auto *N = E->getBase()->getType()->getNominalOrBoundGenericNominal())
     if (auto RV = NominalTypeMemberRefRValueEmitter(E, C, N).emit(SGF))
-      return RValue(std::move(RV).getValue());
+      return RValue(const_cast<RValue&&>(std::move(RV).getValue()));
```
Why is this necessary? An rvalue reference-to-`const` has almost no reason to exist.

hughbe replied:
Hmm, seems like a Visual Studio bug, rather than something MSVC specific. I can reproduce this from the IDE in intellisense but not MSVC. I wrote this in about a month ago so I think I just assumed IDE errors == MSVC errors for this.

The error is:
```
swift::lowering::RValue(const swift::Lowering::RValue &) is inaccessible
```

hughbe replied:
... I've reverted this cahnge

hughbe commented:
@gottesmm FYI

compnerd commented:
@swift-ci please test

hughbe commented:
What's causing these failures - it doesn't look like things are actually failing in the CI?

jrose-apple commented:
"Wrong SHA" failures are a workaround for a bug with the Jenkins GitHub plugin we're using. There are general plans to upgrade the CI system so meanwhile we just put in an expedient hack to fail and automatically retry the build.

hughbe commented:
@jrose-apple thanks for letting me know. I'm never sure if tests passed or not, but seems good

hughbe commented:
Tests seem to have passed, this was never merged

hughbe commented:
Derp, you did a test only  trigger

jrose-apple commented:
Ah, yeah, hoping for final sign-off from @jckarter.

jckarter commented on include/swift/SILOptimizer/PassManager/PassPipeline.h:
```diff
@@ -56,7 +56,9 @@ class SILPassPipelinePlan final {
   ~SILPassPipelinePlan() = default;
   SILPassPipelinePlan(const SILPassPipelinePlan &) = default;
+#if !defined(_MSC_VER) || defined(__clang__)
   SILPassPipelinePlan(SILPassPipelinePlan &&) = delete;
```
If MSVC doesn't like the `= delete`, can we at least leave an `llvm_unreachable` implementation behind? It must have been deleted for a reason, and it'd be nice to still check dynamically.

hughbe replied:
Yeah, this has been fixed BTW: https://connect.microsoft.com/VisualStudio/feedback/details/3116505/msvc-fails-to-compile-code-that-compiles-with-clang-reports-attempting-to-reference-a-deleted-function-error-from-destructor

hughbe replied:
I'm not sure that's the right thing to do.

The errors are
> 1>C:\Users\hbellamy\Documents\GitHub\my-swift\swift\lib\SILOptimizer\PassManager\PassPipeline.cpp(104): error C2280: 'swift::SILPassPipelinePlan::SILPassPipelinePlan(swift::SILPassPipelinePlan &&)': attempting to reference a deleted function
> 1>  C:\Users\hbellamy\Documents\GitHub\my-swift\swift\include\swift/SILOptimizer/PassManager/PassPipeline.h(51): note: see declaration of 'swift::SILPassPipelinePlan::SILPassPipelinePlan'

It comes from [this code](https://github.com/apple/swift/blob/master/lib/SILOptimizer/PassManager/PassPipeline.cpp)

```
SILPassPipelinePlan
SILPassPipelinePlan::getDiagnosticPassPipeline(SILOptions Options) {
  SILPassPipelinePlan P;

  if (SILViewSILGenCFG) {
    addCFGPrinterPipeline(P, "SIL View SILGen CFG");
  }

  // If we are asked do debug serialization, instead of running all diagnostic
  // passes, just run mandatory inlining with dead transparent function cleanup
  // disabled.
  if (Options.DebugSerialization) {
    addMandatoryDebugSerialization(P);
    return P;
  }

  // Lower all ownership instructions right after SILGen for now.
  addOwnershipModelEliminatorPipeline(P);

  // Otherwise run the rest of diagnostics.
  addMandatoryOptPipeline(P);

  if (SILViewGuaranteedCFG) {
    addCFGPrinterPipeline(P, "SIL View Guaranteed CFG");
  }
  return P;
}
```

hughbe replied:
@gottesmm  FYI

jckarter commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -3633,7 +3635,13 @@ void ArgEmitter::emitShuffle(Expr *inner,
+      // extent.params.begin(), extent.params.end())" due to a missing "<"
+      // parameter.
+      for (auto param : extent.Params) {
+        innerParams.append(std::initializer_list<SILParameterInfo>({ param }));
```
`innerParams.push_back(param);` would be simpler.

hughbe replied:
Fixed

jckarter commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -4239,7 +4247,9 @@ namespace {
         // the AST-level abstraction pattern, to ensure that function
         // types in payloads are re-abstracted correctly.
+#if !defined(_MSC_VER) || defined(__clang__)
         assert(!AssumedPlusZeroSelf);
```
We should look into what's making this assertion fail. It could indicate a serious problem.

hughbe replied:
For now, I'm going to revert this change. It's possible this is related to my changes elsewhere in my working branch. However, in previous PRs I have seen assertions happening as they assume a define left to right ordering of method results passed to a method, so could be something more

hughbe commented:
Updated the PR to address your feedback. I also rebased, and fixed several more MSVC errors. Some are known. A new one is https://connect.microsoft.com/VisualStudio/feedback/details/3117167

jckarter commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -101,7 +101,7 @@ SILPassPipelinePlan::getDiagnosticPassPipeline(SILOptions Options) {
   if (Options.DebugSerialization) {
     addMandatoryDebugSerialization(P);
-    return P;
+    return const_cast<const SILPassPipelinePlan&>(P);
```
What is this cast for? This probably defeats NRVO. Looking at the code here, I suspect the deleted move constructor was intended to prevent people from writing pessimizing `std::move`s, but Clang has a warning for that now. AFAICT it should be safe to just remove the `SILPassPipelinePlan(SILPassPipelinePlan &&) = delete;` declaration.

hughbe replied:
I was just experimenting with ideas to fix the error. I can remove the delete decleration if you'd like

jckarter replied:
I don't see any obvious reason to keep it. Pinging @gottesmm since he put it there.

hughbe replied:
Removed, thanks

jckarter commented:
Looks good, thanks @hughbe !

jrose-apple commented:
@swift-ci Please smoke test
